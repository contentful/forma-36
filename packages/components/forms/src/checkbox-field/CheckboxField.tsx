import React from 'react';
import { ControlledInputField } from '../controlled-input-field';
import type { ControlledInputFieldProps } from '../controlled-input-field';

export type CheckboxFieldProps = Omit<ControlledInputFieldProps, 'inputType'>;

export const _CheckboxField = (
  { testId = 'cf-ui-checkbox-field', ...otherProps }: CheckboxFieldProps,
  ref: React.Ref<HTMLDivElement>,
) => {
  return (
    <ControlledInputField
      testId={testId}
      {...otherProps}
      ref={ref}
      inputType="checkbox"
    />
  );
};

export const CheckboxField = React.forwardRef(_CheckboxField);
