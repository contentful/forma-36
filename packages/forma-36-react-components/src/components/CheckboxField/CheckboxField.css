@import 'settings/typography';

.CheckboxField {
  display: inline-flex;
  align-items: baseline;
}

.CheckboxField--disabled {
  opacity: 0.7;
}

.CheckboxField__checkbox {
  margin-right: calc(1rem * (8 / var(--font-base-default)));
  transform: translateY(1px);
}

.CheckboxField__label {
  margin-bottom: 0;
}

.CheckboxField__label--light {
  font-weight: var(--font-weight-normal);
}
