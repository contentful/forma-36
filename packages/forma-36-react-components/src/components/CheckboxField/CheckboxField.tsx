import React from 'react';
import ControlledInputField, {
  ControlledInputFieldPropTypes,
} from '../ControlledInputField';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface CheckboxFieldProps extends ControlledInputFieldPropTypes {}

const defaultProps: Partial<CheckboxFieldProps> = {
  labelIsLight: false,
  checked: false,
  inputType: 'checkbox',
  testId: 'cf-ui-checkbox-field',
};

export const CheckboxField = (props: CheckboxFieldProps) => {
  const { testId, ...otherProps } = props;

  return (
    <ControlledInputField
      testId={testId}
      {...otherProps}
      inputType="checkbox"
    />
  );
};

CheckboxField.defaultProps = defaultProps;

export default CheckboxField;
