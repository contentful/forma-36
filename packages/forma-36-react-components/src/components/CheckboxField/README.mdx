Use CheckboxField to give users the possibility to select multiple elements, zero, one or more items. This component can be used for example in forms or terms and conditions. It contains a checkbox with its label. For a single choice, please consider using RadioButton component.

## Table of contents

- [How to use CheckboxField](#how-to-use-checkboxField)
- [Code examples](#code-examples)
- [Content recommendations](#content-recommendations)
- [Accessibility](#accessibility)

## How to use CheckboxField

- Checkboxes should work independently, checking one should not affect any other checkbox in the interface
- Checkbox should always have a label
- Listing options for a user should have some order. Some sort of system is most likely expected by the user when working with lists (numerical, time based)
- For the label, 2 different styles can be used, depending on the context and importance one wants to give to the checkbox. The default is the medium weight label. You can change it to a regular weight label by using `labelIsLight` property

NOTE: When attaching a help text to a checkbox, the medium label has to be used to ensure that there's enough contrast.

## Code examples

Basic usage

```jsx
// import {CheckboxField} from '@contentful/forma-36-react-components';

<CheckboxField labelText="Label text" />
```

Example using helper text, which will be displayed below the label

```jsx
// import {CheckboxField} from '@contentful/forma-36-react-components';

<CheckboxField
  name="checkbox name"
  id="randomId"
  labelText="Label text"
  helpText="Helper text displayed below"
/>
```

Light label example - `labelIsLight` changes label font weight to regular. Please keep in mind to not use it with helper text. For that case the medium label has to be used to ensure that there's enough contrast.

```jsx
// import {CheckboxField} from '@contentful/forma-36-react-components';

<CheckboxField
  name="checkbox name"
  id="randomId"
  labelText="Label text"
  labelIsLight
/>
```

## Content recommendations

- Start every option from a capital letter
- Do not use commas at the end of each line
- If the checkbox is used to agree to terms and conditions, we recommend to use first person in the sentence

## Accessibility

- Checkboxes should have a clear and concise label.
- Users should be warned if selecting a checkbox will cause a change in context.
- When the checkbox has focus, pressing the Space key changes the state of the checkbox.
