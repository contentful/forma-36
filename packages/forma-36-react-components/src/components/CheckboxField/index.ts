export { CheckboxField } from './CheckboxField';
export type { CheckboxFieldProps } from './CheckboxField';
