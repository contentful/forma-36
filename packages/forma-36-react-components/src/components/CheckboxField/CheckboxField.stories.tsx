import React, { useState } from 'react';

import FieldGroup from '../Form/FieldGroup';
import CheckboxField, { CheckboxFieldProps } from './CheckboxField';
import Flex from '../Flex/Flex';
import SectionHeading from '../Typography/SectionHeading';

export default {
  title: 'Form Elements/CheckboxField',
  component: CheckboxField,
  parameters: {
    propTypes: [CheckboxField['__docgenInfo']],
  },
  argTypes: {
    className: { control: { disable: true } },
    testId: { control: { disable: true } },
  },
};

export const Basic = (args: CheckboxFieldProps) => {
  const [optionOne, setOptionOne] = useState(false);
  const [optionTwo, setOptionTwo] = useState(false);

  return (
    <FieldGroup>
      <CheckboxField
        {...args}
        id="Checkbox1"
        checked={optionOne}
        value="yes"
        onChange={(e) => setOptionOne((e.target as HTMLInputElement).checked)}
      />
      <CheckboxField
        {...args}
        id="Checkbox2"
        value="yes"
        checked={optionTwo}
        onChange={(e) => setOptionTwo((e.target as HTMLInputElement).checked)}
      />
    </FieldGroup>
  );
};

Basic.args = {
  labelText: 'some label text',
  name: 'some name',
  validationMessage: 'validation message',
  helpText: 'help text',
};

export const overview = () => (
  <>
    <Flex marginBottom="spacingS" marginTop="spacingM">
      <SectionHeading element="h3">Checkbox Field default</SectionHeading>
    </Flex>
    <CheckboxField
      labelText="Option 1"
      helpText="This is a helptext"
      name="someOption"
      checked
      value="yes"
      id="termsCheckbox"
    />
    <Flex marginBottom="spacingS" marginTop="spacingM">
      <SectionHeading element="h3">
        Checkbox Field with validation message
      </SectionHeading>
    </Flex>
    <CheckboxField
      labelText="Option 2"
      helpText="This is a helptext"
      validationMessage="validationMessage"
      name="someOption"
      value="no"
      id="termsCheckboxOption2"
    />
    <Flex marginBottom="spacingS" marginTop="spacingM">
      <SectionHeading element="h3">Checkbox Field diabled</SectionHeading>
    </Flex>
    <CheckboxField
      labelText="Option 2"
      helpText="This is a helptext"
      disabled
      name="someOption"
      value="no"
      id="termsCheckboxOption3"
    />
    <Flex marginBottom="spacingS" marginTop="spacingM">
      <SectionHeading element="h3">
        Checkbox Field diabled checked
      </SectionHeading>
    </Flex>
    <CheckboxField
      labelText="Option 2"
      helpText="This is a helptext"
      disabled
      checked
      name="someOption"
      value="no"
      id="termsCheckboxOption4"
    />
  </>
);
