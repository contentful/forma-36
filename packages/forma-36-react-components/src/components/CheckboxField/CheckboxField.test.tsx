import React from 'react';
import { shallow, mount } from 'enzyme';
import axe from '../../utils/axeHelper';
import CheckboxField from './CheckboxField';

it('renders the component', () => {
  const output = shallow(
    <CheckboxField id="checkbox" labelText="label text" />,
  );

  expect(output).toMatchSnapshot();
});

it('has no a11y issues', async () => {
  const output = mount(
    <CheckboxField id="checkbox" labelText="label text" />,
  ).html();
  const results = await axe(output);

  expect(results).toHaveNoViolations();
});
