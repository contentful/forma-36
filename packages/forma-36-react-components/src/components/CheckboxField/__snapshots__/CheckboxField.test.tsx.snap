// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component 1`] = `
.emotion-4 {
  margin-bottom: 0;
}

.emotion-5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  display: inline-block;
  font-size: 0.875rem;
  opacity: 0;
  position: absolute;
  margin: 0;
  width: 1rem;
  height: 1rem;
  z-index: 1;
  cursor: pointer;
}

.emotion-0:focus + label {
  outline: none;
  border: 2px solid #2E75D4;
  box-shadow: 0px 0px 0px 3px #84B9F5;
}

.emotion-0:checked + label {
  border-color: #2E75D4;
  background: #2E75D4;
}

.emotion-0:disabled + label {
  border-color: #c3cfd5;
  background: #c3cfd5;
}

.emotion-0:checked + label,
.emotion-0:indeterminate + label {
  border-color: #2E75D4;
  background: #2E75D4;
}

.emotion-0:checked:disabled + label svg,
.emotion-0:indeterminate:disabled + label svg {
  fill: #6B7888;
}

.emotion-0:disabled + label {
  border-color: #c3cfd5;
  background: #c3cfd5;
}

.emotion-0:disabled + label svg {
  fill: #c3cfd5;
}

.emotion-2 {
  background: #ffffff;
  width: 1rem;
  height: 1rem;
  border: 2px solid #c3cfd5;
  fill: #ffffff;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  box-sizing: border-box;
  border-radius: 20%;
}

.emotion-2 svg {
  width: 0.75rem;
  min-width: 0.75rem;
  min-height: 0.75rem;
  height: 0.75rem;
}

.emotion-3 {
  display: inline-block;
  position: relative;
  text-overflow: ellipsis;
  line-height: 1.5rem;
  margin: 0;
  width: 1rem;
  height: 1rem;
  min-height: 1rem;
  overflow: visible;
  margin-right: 0.5rem;
  margin-top: 0.25rem;
}

.emotion-1 {
  display: inline-block;
  fill: #ffffff;
  height: 24px;
  width: 24px;
}

<div
  class="emotion-5"
  data-test-id="cf-ui-checkbox-field"
>
  <div
    class="emotion-3"
    data-test-id="cf-ui-controlled-input"
  >
    <input
      aria-label="label text"
      class="emotion-0"
      id="checkbox"
      type="checkbox"
      value=""
    />
    <label
      class="emotion-2"
      for="checkbox"
    >
      <svg
        class="emotion-1"
        data-test-id="cf-ui-icon"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M20.7071 5.29289C21.0976 5.68342 21.0976 6.31658 20.7071 6.70711L9.70711 17.7071C9.51957 17.8946 9.26522 18 9 18C8.73478 18 8.48043 17.8946 8.29289 17.7071L3.29289 12.7071C2.90237 12.3166 2.90237 11.6834 3.29289 11.2929C3.68342 10.9024 4.31658 10.9024 4.70711 11.2929L9 15.5858L19.2929 5.29289C19.6834 4.90237 20.3166 4.90237 20.7071 5.29289Z"
        />
      </svg>
    </label>
  </div>
  <div>
    <label
      class="FormLabel__FormLabel___3d6zQ emotion-4"
      data-test-id="cf-ui-form-label"
      for="checkbox"
    >
      label text
    </label>
  </div>
</div>
`;
